src/core/qgstessellator.cpp

```diff
@@ -473,337 +543,275 @@ double _minimum_distance_between_coordinates( const QgsPolygon &polygon )
+  const QgsPoint *originOffset,
+  bool reverse
+)
+{
+  // if reverse is true, the triangle vertices are added in reverse order and normal is inverted
+  const QVector3D normal = reverse ? -pNormal : pNormal;
+  for ( int i = 0; i < 3; ++i )
+  {
+    const int index = reverse ? 2 - i : i;
+
+    // cppcheck-suppress negativeContainerIndex
+    QVector3D point = points[ index ];
+    const float z = mNoZ ? 0.0f : point.z();
+    QVector4D pt( point.x(), point.y(), z, 0 );
+
+    pt = *transformMatrix * pt;
+
+    const double fx = pt.x() - mOrigin.x() + originOffset->x();
+    const double fy = pt.y() - mOrigin.y() + originOffset->y();
+    const double baseHeight = mNoZ ? 0 : pt.z() + originOffset->z();
+    const double fz = mNoZ ? 0.0 : ( baseHeight + extrusionHeight );
+
+    if ( baseHeight < mZMin )
+      mZMin =  static_cast<float>( baseHeight );
+    if ( baseHeight > mZMax )
+      mZMax = static_cast<float>( baseHeight );
+    if ( fz > mZMax )
+      mZMax = static_cast<float>( fz );
+
+    // NOLINTBEGIN(bugprone-branch-clone)
+    if ( mOutputZUp )
+    {
+      mData << static_cast<float>( fx ) << static_cast<float>( fy ) << static_cast<float>( fz );
+      if ( mAddNormals )
+        mData << normal.x() << normal.y() << normal.z();
+    }
+    else
+    {
+      mData << static_cast<float>( fx ) << static_cast<float>( fz ) << static_cast<float>( -fy );
+      if ( mAddNormals )
+        mData << normal.x() << normal.z() << - normal.y();
+    }
+    // NOLINTEND(bugprone-branch-clone)
+
+    if ( mAddTextureCoords )
+    {
+      const std::pair<float, float> pr = rotateCoords( static_cast<float>( pt.x() ), static_cast<float>( pt.y() ), 0.0f, 0.0f, mTextureRotation );
+      mData << pr.first << pr.second;
+    }
+  }
+}
+
+std::vector<QVector3D> QgsTessellator::generateConstrainedDelaunayTriangles( const QgsPolygon *polygonNew )
+{
+  QList<std::vector<p2t::Point *>> polylinesToDelete;
+  QHash<p2t::Point *, float> z;
+
+  // polygon exterior
+  std::vector<p2t::Point *> polyline;
+  _ringToPoly2tri( qgsgeometry_cast< const QgsLineString * >( polygonNew->exteriorRing() ), polyline, mNoZ ? nullptr : &z );
+  polylinesToDelete << polyline;
+
+  p2t::CDT cdt = p2t::CDT( polyline );
+
+  // polygon holes
+  for ( int i = 0; i < polygonNew->numInteriorRings(); ++i )
+  {
+    std::vector<p2t::Point *> holePolyline;
+    const QgsLineString *hole = qgsgeometry_cast< const QgsLineString *>( polygonNew->interiorRing( i ) );
+
+    _ringToPoly2tri( hole, holePolyline, mNoZ ? nullptr : &z );
+
+    cdt.AddHole( holePolyline );
+    polylinesToDelete << holePolyline;
+  }
+
+  cdt.Triangulate();
+  std::vector<p2t::Triangle *> triangles = cdt.GetTriangles();
+
+  std::vector<QVector3D> trianglePoints;
+  trianglePoints.reserve( triangles.size() * 3 );
+
+  for ( p2t::Triangle *t : triangles )
+  {
+    trianglePoints.emplace_back( QVector3D( t->GetPoint( 0 )->x / mScale, t->GetPoint( 0 )->y / mScale, z.value( t->GetPoint( 0 ) ) ) );
```

Comment:
With `emplace_back`, shouldn't these be:
```suggestion
    trianglePoints.emplace_back( t->GetPoint( 0 )->x / mScale, t->GetPoint( 0 )->y / mScale, z.value( t->GetPoint( 0 ) ) );
```
to avoid the copy?